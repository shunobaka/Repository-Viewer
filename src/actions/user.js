/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update user state.
 */
import { USERS_SEARCH_RESULT, USER_LOADED } from './types';

/**
 * Returns a redux action object used to save users and username query in state
 *    based on the provided users and query.
 * @param {array} users User objects to be saved in the redux state
 * @param {*} query The username query by which the users were retrieved
 * @returns Redux action object
 * @export
 */
export const usersLoaded = (users, query) => {
  return {
    type: USERS_SEARCH_RESULT,
    payload: {
      users,
      query,
    },
  };
};

/**
 * Returns a redux action object used to save user in state based on the provided
 *    user.
 * @param {object} user User object to be saved in redux state
 * @returns Redux action object
 * @export
 */
export const userLoaded = (user) => {
  return {
    type: USER_LOADED,
    payload: user,
  };
};
