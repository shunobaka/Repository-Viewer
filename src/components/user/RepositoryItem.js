import React from 'react';
import { Accordion, Card } from 'react-bootstrap';
import PropTypes from 'prop-types';

const RepositoryItem = ({ repository, index }) => {
  return (
    <Card>
      <Accordion.Toggle as={Card.Header} eventKey={repository.id}>
        {`${index + 1}.`.padEnd(4)} {repository.name}
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={repository.id}>
        <Card.Body>
          <a href={repository.url}>Go to repository</a>
          <br />
          <br />
          <h5>Description:</h5>
          {repository.description ? (
            <span>{repository.description}</span>
          ) : (
            <span className="text-warning">
              This repository does not have a description
            </span>
          )}
          <br />
          <br />
          <h5>Created on:</h5>
          <span>{new Date(repository.created_at).toLocaleDateString()}</span>
        </Card.Body>
      </Accordion.Collapse>
    </Card>
  );
};

RepositoryItem.propTypes = {
  repository: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    url: PropTypes.string,
    description: PropTypes.string,
    created_at: PropTypes.string,
  }),
  index: PropTypes.number.isRequired,
};

export default RepositoryItem;
