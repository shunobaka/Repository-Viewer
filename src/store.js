/**
 * @fileoverview Creates and exports the redux store.
 */
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import combinedReducer from './reducers';

const middleware = [thunk];

/**
 * The redux store that is created and exported.
 * @export
 */
const store = createStore(
  combinedReducer,
  composeWithDevTools(applyMiddleware(...middleware))
);

export default store;
