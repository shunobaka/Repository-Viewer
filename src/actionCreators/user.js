import githubApi from '../utils/githubApi';
import { usersLoaded, userLoaded } from '../actions/user';
import { setAlert } from '../actionCreators/alert';

const getUsers = (nameQuery) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/search/users?q=${nameQuery}`);

    const payload = res.data.items.map((item) => ({
      username: item.login,
      avatar: item.avatar_url,
      id: item.id,
      num_repos: item.public_repos,
    }));

    dispatch(usersLoaded(payload, nameQuery));
  } catch (err) {
    dispatch(
      setAlert(
        'There was a problem searching for users, please try again.',
        false
      )
    );
  }
};

const loadUser = (username) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/users/${username}`);

    const user = {
      id: res.data.id,
      username: res.data.login,
      avatar: res.data.avatar_url,
      num_repos: res.data.public_repos,
    };

    dispatch(userLoaded(user));
  } catch (err) {
    dispatch(setAlert('Could not retrieve user, please try again.', false));
  }
};

export default {
  getUsers,
  loadUser,
};
