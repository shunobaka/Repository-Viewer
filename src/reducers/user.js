import { USER_LOADED, USERS_SEARCH_RESULT } from '../actions/types';

const initialState = {
  user: null,
  users: [],
  query: null,
  loading: true,
};

export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case USER_LOADED:
      return {
        ...state,
        user: payload,
      };
    case USERS_SEARCH_RESULT:
      return {
        user: null,
        users: payload.users,
        query: payload.query,
        loading: false,
      };
    default:
      return state;
  }
}
