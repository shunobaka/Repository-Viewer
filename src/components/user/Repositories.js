/**
 * @fileoverview Defines a Repositories react component that renders a list of
 *    repositories.
 */
import React from 'react';
import { Container, Row, Col, Accordion } from 'react-bootstrap';
import PropTypes from 'prop-types';
import RepositoryItem from './RepositoryItem';

/**
 * Repositories react component that renders a list of repositories.
 * @param {object} props Contains the array of repositories to be displayed
 */
const Repositories = ({ repositories }) => {
  return (
    <Container>
      <Row>
        <Col lg={{ span: 6, offset: 3 }}>
          {repositories.length > 0 ? (
            <Accordion>
              {repositories.map((repo, index) => {
                return (
                  <RepositoryItem
                    key={repo.id}
                    index={index}
                    repository={repo}
                  />
                );
              })}
            </Accordion>
          ) : (
            <div className="alert alert-warning">
              No repositories match the filter!
            </div>
          )}
        </Col>
      </Row>
    </Container>
  );
};

Repositories.propTypes = {
  repositories: PropTypes.array.isRequired,
};

export default Repositories;
