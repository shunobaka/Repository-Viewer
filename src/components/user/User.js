import React, { Fragment, useEffect, useState } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import repositoryActionCreator from '../../actionCreators/repository';
import userActionCreator from '../../actionCreators/user';
import {
  Form,
  FormControl,
  Button,
  Container,
  Row,
  Col,
} from 'react-bootstrap';
import Repositories from './Repositories';

const User = ({
  user,
  repositories,
  actions: { getRepositoriesForUser, filterRepositories, loadUser },
  match,
}) => {
  const [filterInput, setFilterInput] = useState('');

  useEffect(() => {
    loadUser(match.params.username);
    getRepositoriesForUser(match.params.username);
  }, [loadUser, getRepositoriesForUser, match.params.username]);

  const onChange = (e) => {
    setFilterInput(e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();

    filterRepositories(filterInput);
  };

  return (
    <Fragment>
      {user ? (
        <Fragment>
          <h2 className="text-center green-text">
            Viewing repositories for {user.username}
          </h2>
          {user.num_repos > 0 ? (
            <Fragment>
              <h4 className=" text-center green-text">
                Click repository name to view details
              </h4>
              <br />
              <Container>
                <Row className="justify-content-md-center">
                  <Col lg="auto">
                    <h4 className="text-center green-text">
                      Filter repositories by name:
                    </h4>
                    <Form inline onSubmit={onSubmit}>
                      <FormControl
                        type="text"
                        className="mr-sm-2 ml-auto"
                        value={filterInput}
                        onChange={onChange}
                      />
                      <Button type="submit" variant="outline-success mr-auto">
                        Filter
                      </Button>
                    </Form>
                  </Col>
                </Row>
              </Container>
              <br />
              <Repositories repositories={repositories} />
            </Fragment>
          ) : (
            <h4>User has no repositories</h4>
          )}
        </Fragment>
      ) : (
        <h2>User loading</h2>
      )}
    </Fragment>
  );
};

User.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    num_repos: PropTypes.number.isRequired,
  }),
  repositories: PropTypes.array.isRequired,
  actions: PropTypes.shape({
    getRepositoriesForUser: PropTypes.func.isRequired,
    filterRepositories: PropTypes.func.isRequired,
    loadUser: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = (state) => ({
  user: state.user.user,
  repositories: state.repository.displayed_repositories,
});

const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(
      { ...repositoryActionCreator, ...userActionCreator },
      dispatch
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(User);
