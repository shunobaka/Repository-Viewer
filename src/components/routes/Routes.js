/**
 * @fileoverview Defines a Routes react component that renders a react router.
 */
import React from 'react';
import { Route, Switch } from 'react-router-dom';

import Landing from '../../pages/Landing/Landing';
import NotFound from '../../pages/NotFound/NotFound';
import Search from '../../pages/Search/Search';
import User from '../../pages/User/User';

/**
 * Routes react component that renders a react router specifying accessible routes.
 */
const Routes = () => {
  return (
    <Switch>
      <Route exact path="/" component={Landing} />
      <Route exact path="/search/:query" component={Search} />
      <Route exact path="/user/:username" component={User} />
      <Route component={NotFound} />
    </Switch>
  );
};

export default Routes;
