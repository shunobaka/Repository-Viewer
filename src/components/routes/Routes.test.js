/**
 * @fileoverview Defines snapshot unit tests for Routes component.
 */
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import renderer, { act } from 'react-test-renderer';
import Routes from './Routes';
import store from '../../store';
import { usersLoaded, userLoaded } from '../../actions/user';
import userActionCreators from '../../actionCreators/user';
import repositoryActionCreators from '../../actionCreators/repository';

/** Mock functions to prevent creating http requests */
jest.mock('../../actionCreators/repository.js');
jest.mock('../../actionCreators/user.js');

/** Mock the actionCreators used by the component to do nothing */
userActionCreators.loadUser.mockImplementation(() => {
  return { type: 'mock_action' };
});

userActionCreators.getUsers.mockImplementation(() => {
  return { type: 'mock_action' };
});

repositoryActionCreators.getRepositoriesForUser.mockImplementation(() => {
  return { type: 'mock_action' };
});

describe('routes snapshots', () => {
  let history;

  beforeEach(() => {
    history = createMemoryHistory();
  });

  afterEach(() => {
    history = null;
  });

  it('home route renders correctly', () => {
    history.push('/');

    const tree = renderer
      .create(
        <Router history={history}>
          <Routes />
        </Router>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('search route renders correctly', () => {
    history.push('/search/test');

    act(() => {
      store.dispatch(usersLoaded([], ''));
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <Routes />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('user route renders correctly', () => {
    history.push('/user/test');

    const mockUser = {
      username: 'Test',
      num_repos: 0,
    };

    act(() => {
      store.dispatch(userLoaded(mockUser));
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <Routes />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('unknown route renders correctly', () => {
    history.push('/not-exist');

    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <Routes />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
