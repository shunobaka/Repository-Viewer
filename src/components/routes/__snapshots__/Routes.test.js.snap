// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes snapshots home route renders correctly 1`] = `
<section
  className="landing"
>
  <div
    className="landing-inner"
  >
    <h1>
      Welcome to Github-Repo-Viewer
    </h1>
    <h3>
      Search for a user in the navigation above to view repositories.
    </h3>
  </div>
</section>
`;

exports[`routes snapshots search route renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Search results for 
    
    :
  </h2>,
  <div
    className="text-center container"
  >
    <br />
    <div
      className="alert alert-warning"
    >
      No users found matching 
      
      !
    </div>
  </div>,
]
`;

exports[`routes snapshots unknown route renders correctly 1`] = `
<section
  className="landing"
>
  <div
    className="landing-inner"
  >
    <h1>
      The requested page was not found.
    </h1>
    <h3>
      Please try again.
    </h3>
  </div>
</section>
`;

exports[`routes snapshots user route renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Viewing repositories for 
    Test
  </h2>,
  <div
    className="text-center container"
  >
    <br />
    <div
      className="alert alert-warning"
    >
      User has no repositories!
    </div>
  </div>,
]
`;
